File: src/main/java/pl/sparkbit/commons/logging/OutputSizeCappingFilter.java
package pl.sparkbit.commons.logging;

import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.Logger;
import ch.qos.logback.classic.turbo.TurboFilter;
import ch.qos.logback.core.spi.FilterReply;
import ch.qos.logback.core.status.WarnStatus;
import org.slf4j.Marker;

import java.util.Arrays;
import java.util.concurrent.atomic.AtomicLong;


/**
 * Logback [Filter] which counts the output size of all messages, and beyond [maxBytes], drops new messages.
 * The counter is restarted every [epochLengthMillis].
 */
public class OutputSizeCappingFilter extends TurboFilter {

    private int maxBytes = 60_000_000;
    private long epochLengthMillis = 60_000;
    private EpochBytesCounter latestEpochCounter = new EpochBytesCounter(0);

    @Override
    public void start() {
        long epochIdentifier = System.currentTimeMillis() / epochLengthMillis;
        synchronized (this) {
            latestEpochCounter = new EpochBytesCounter(epochIdentifier);
        }

        super.start();
    }

    @Override
    public void stop() {
        synchronized (this) {
            latestEpochCounter = null;
        }

        super.stop();
    }

    private EpochBytesCounter currentEpoch() {
        long newEpochIdentifier = System.currentTimeMillis() / epochLengthMillis;
        synchronized (this) {
            if (newEpochIdentifier > latestEpochCounter.epochIdentifier) {
                latestEpochCounter = new EpochBytesCounter(newEpochIdentifier);
            }
            return latestEpochCounter;
        }
    }

    private final class EpochBytesCounter {
        private final long epochIdentifier;
        private AtomicLong bytesCounter;
        private boolean reporterd = false;

        private EpochBytesCounter(long epochIdentifier) {
            this.epochIdentifier = epochIdentifier;
            this.bytesCounter = new AtomicLong(0L);
        }

        public long incrementCounter(long newBytes) {
            return bytesCounter.addAndGet(newBytes);
        }

        public void reportOnce() {
            synchronized (this) {
                if (reporterd) {
                    return;
                }
                reporterd = true;
            }

            addStatus(
                    new WarnStatus(
                            "Log size limit reached - " + maxBytes +
                                    " bytes. New messages wil not be appended for a while.",
                            OutputSizeCappingFilter.this
                    )
            );
        }
    }

    @Override
    public FilterReply decide(final Marker marker, final Logger logger, final Level level,
                              final String format, final Object[] params, final Throwable t) {
        EpochBytesCounter epoch = currentEpoch();

        long newBytesEstimate = logger.getName().length() * 2
                + ((level != null) ? level.toString().length() * 2 : 0)
                + ((format != null) ? format.length() : 0)
                + ((params != null) ? Arrays.stream(params)
                    .filter(s -> s != null)
                    .mapToInt(o -> o.toString().length())
                    .sum() : 0)
                + Arrays.stream(t.getStackTrace()).filter(s -> s != null).mapToInt(o -> o.toString().length()).sum();
        long totalSize = epoch.incrementCounter(newBytesEstimate);
        if (totalSize > maxBytes) {
            epoch.reportOnce();
            return FilterReply.DENY;
        }

        return FilterReply.NEUTRAL;
    }
}

File: src/main/kotlin/pl/sparkbit/commons/logging/OutputSizeCappingFilter.kt
package pl.sparkbit.commons.logging

import ch.qos.logback.classic.spi.ILoggingEvent
import ch.qos.logback.core.filter.Filter
import ch.qos.logback.core.spi.FilterReply
import ch.qos.logback.core.status.WarnStatus
import java.util.concurrent.atomic.AtomicLong

/**
 * Logback [Filter] which counts the output size of all messages, and beyond [maxBytes], drops new messages.
 * The counter is restarted every [epochLengthMillis].
 */
@Suppress("unused")
class OutputSizeCappingFilter : Filter<ILoggingEvent>() {
    var maxBytes: Long = 60_000_000
    var epochLengthMillis: Long = 60_000
    private var latestEpochCounter: EpochBytesCounter = EpochBytesCounter(0)

    private inner class EpochBytesCounter(
        val epochIdentifier: Long
    ) {
        private val bytesCounter: AtomicLong = AtomicLong(0)
        private var reported = false

        fun incrementCounter(newBytes: Long): Long {
            return bytesCounter.addAndGet(newBytes)
        }

        fun reportOnce() {
            synchronized(this) {
                if (reported) {
                    return
                }
                reported = true
            }
            context.statusManager
                .add(
                    WarnStatus(
                        "Log size limit reached  - $maxBytes bytes. New messages will not be appended for a while.",
                        this@OutputSizeCappingFilter
                    )
                )
        }
    }

    override fun decide(event: ILoggingEvent): FilterReply {
        val epoch = currentEpoch()

        val newBytes = event.estimateMessageSizeInBytes()

        val totalSize = epoch.incrementCounter(newBytes)
        return if (totalSize > maxBytes) {
            epoch.reportOnce()
            FilterReply.DENY
        } else {
            FilterReply.NEUTRAL
        }
    }

    private fun currentEpoch(): EpochBytesCounter {
        val newEpochIdentifier = System.currentTimeMillis() / epochLengthMillis
        synchronized(this) {
            if (newEpochIdentifier > latestEpochCounter.epochIdentifier) {
                latestEpochCounter = EpochBytesCounter(newEpochIdentifier)
            }
            return latestEpochCounter
        }
    }

    private fun ILoggingEvent.estimateMessageSizeInBytes(): Long {
        var sum: Long = 0

        message?.let { sum += it.length * 2 }
        argumentArray?.let { sum += it.size * 50 }
        loggerName?.let { sum += it.length * 2 }

        var t = throwableProxy
        while (t != null) {
            sum += t.className.length * 2
            t.message?.let { sum += it.length * 2 }
            t.stackTraceElementProxyArray?.let { sum += it.size * 80 }
            t = t.cause
        }

        return sum
    }

    override fun toString(): String {
        return "OutputSizeCappingFilter$name"
    }
}
